### ISSUE TITLE
Bug: [no-unnecessary-condition] False positive with nested optional chains if nullability is inferred via type parameters

### ISSUE THREAD

**Issue by cairomassimo** [typescript-eslint/typescript-eslint#8652]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.4.2&fileType=.ts&code=CYUwxgNghgTiAEYD2A7AzgF3gDwFzwAoAeAFQD4C94SBKeAXjOroB94UBXCCAbgCg%2B2AISUaAfgB0owZNGzsNPkA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A&tokens=false

### Repro Code

```TypeScript
declare const x: (<T>(x: T) => T) | null;

x!(x)?.(x)
x?.(x)?.(x)
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true
  }
}
```

### Expected Result

_No errors should be reported._

According to the types specified:
- the expression `x!(x)` can be `null`,
- the expression `x?.(x)` can be `null` or `undefined` (in the latter case, causing optional chaining short-circuit).

In both `x!(x)?.(x)` and `x?.(x)?.(x)`, the first part of the expression can be `null` (and not trigger short-circuit), so the second optional chaining operator is not redundant in either case.

### Actual Result

A spurious error "Unnecessary optional chain on a non-nullish value." is reported on the second optional chain in the expression `x?.(x)?.(x)` (but not in `x!(x)?.(x)` - which should be indistinguishable).

### Additional Info

I encountered this bug while developing a TypeScript utility that allows writing, say, `x.$(f).$(g)` instead `g(f(x))`, which in turn allows more ergonomic use of optional-chaining operators. This issue is relevant for my use-case.

**Comment by bradzacher** [comment]
Do you have a more practical example of this code pattern? Just trying to understand the usecase better.

**Comment by cairomassimo** [comment]
Perhaps a more convincing repro: https://typescript-eslint.io/play/#ts=5.4.2&fileType=.ts&code=CYUwxgNghgTiAEYD2A7AzgF3gDwFzwAoAeAFQD4C94SBKeAXjOroB94UBXCCAbgChQkWAmTosAT3zFylfLQZN5bTt359sAfgB0BcTV0112g8b18gA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A&tokens=false

Here we have two similar expressions, one with optional chaining and one without.
TypeScript rejects one and `no-unnecessary-condition` rejects the other.

**Comment by auvred** [comment]
It looks like the current rule implementation just iterates over the function signatures:
https://github.com/typescript-eslint/typescript-eslint/blob/fc08f969e255b55ea35091b297b358bc434a76eb/packages/eslint-plugin/src/rules/no-unnecessary-condition.ts#L623-L628

So it doesn't know that `T` type in the signature may be nullable

```ts
declare const x: (<T>(x: T) => T) | null;
declare const y: (() => void) | null

x?.(y)?.()
```

[playground](https://typescript-eslint.io/play/#ts=5.4.2&fileType=.ts&code=CYUwxgNghgTiAEYD2A7AzgF3gDwFzwAoAeAFQD4C94SBKeAXjOroB94UBXCCAbgChQkWAmTosAT3wECdRvABuSAJbBW7LhD59sAfgB0BcTX0y%2BQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A&tokens=false)

**Comment by cairomassimo** [comment]
> So it doesn't know that T type in the signature may be nullable

Perhaps a quick fix is treating all types as nullable unless proven not nullable (for the purpose of this rule only)?

Of course even better would be to fully rely on TypeScript to infer the correct type of the callee object if possible.

**Comment by bradzacher** [comment]
> Perhaps a quick fix is treating all types as nullable unless proven not nullable

That is unfortunately not how the types API works. Null and undefined are separate types - not properties of a type.

So to say "this type is nullable" means that the type is a union type including null and or undefined. The same as it is in normal TS.

Generic types are a bit different - they report to us as just a named type. We need to add specific handling for them because you need to inspect their flags to determine they're a generic and then you can resolve the contextual constrains to determine if they are nullable.

**Comment by cairomassimo** [comment]
> Generic types are a bit different - they report to us as just a named type.

Now I get it, thanks! I should have said: "treat named types as if they were nullable", but I just had a look at the code at it doesn't look easy to do...

I'll give this issue a go. If I can come up with a fix today I'll open a PR.

I'll try first getting the right type from TypeScript if possible. If not, I'll implement a custom utility `isPossiblyNullishOrUnknown` that will bail on named types and return `true`, and use that one instead of `isNullableType`.

Thanks!

**Comment by cairomassimo** [comment]
I found a quick way to improve the correctness (PR #8662), but it's still not complete.

It's kind-of a workaround, but I think it's worth merging because it adds some new tests (that would have failed before), including a commented-out one that should pass but doesn't, and it does not complicate the code at all.

However the right solution is to get the right type from TypeScript (narrowed down after optional chaining) and test for `undefined` as well, but I have no idea how to do it.

**Comment by cairomassimo** [comment]
Quick update: there seems to be no way to tell whether `undefined` type is short-circuited out or not. Looking at the types, these two expressions are indistinguishable from one another ([Playground link](https://typescript-eslint.io/play/#ts=5.4.2&showAST=types&fileType=.ts&code=CYUwxgNghgTiAEYD2A7AzgF3gDwFzwAoAeAFQD4C94SBKeAXjOroB94UBXCCAbgCg%2BoSLATJ0WAJ4BGfAQJ1G8AG5IAlsBqDw0OIlSZ4EgEyz5DJivWt4HFKABmqlCGADsAfgB0BaTS-y%2BD29jP29NIA&eslintrc=&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A&tokens=true)):

```ts
declare const x: (<T>(x: T) => T) | null;

declare const y1: (() => void)
declare const y2: (() => void) | undefined

x?.(y1)?.()
x?.(y2)?.()
```

I don't think there is a way to implement the correct behavior (i.e., reject the first, allow the second) with the current way typescript-eslint gets the types from typescript.

(One could argue that's a typescript Compiler API bug: the sub-expression `x?.(y1)`, if part of a longer optional chain, can never actually evaluate to `undefined` thanks to short-circuiting - only the full optional chain could ever do. If I have time I'll investigate this more and report it to the TS team.)

**Comment by JoshuaKGoldberg** [comment]
> I don't think there is a way to implement the correct behavior

...yeah I'm not convinced this is something we can really action on well. Type parameters can be falsy/nullable/etc unless constrained otherwise. It's a whole mess.

**Comment by JoshuaKGoldberg** [comment]
👋 just checking in @cairomassimo, is this something you'd like us to keep open?

**Comment by JoshuaKGoldberg** [comment]
It's been a week and we're in the process of cutting through our issue and PR queues. Closing this issue out. But if anybody comes back to this and wants to continue the conversation, please do! If this issue is locked by the time you want to post, feel free to make a new one that links to it + adds new info. Cheers! 🙂

**Comment by cairomassimo** [comment]
Got it, thanks @JoshuaKGoldberg!
